### PR TITLE
:package: serialization support +  :level_slider: update OS & python version in test-matrix

### LINKED ISSUE #85: add windows and mac as platforms to out testing matrix
Should be roughly the same as we did with the `tsflex` packages @jvdd.

### PR DESCRIPTION
**Pull request by jvdd**
Should fix #85 

In shot this PR adds the following contributions;
* Add serialization to plotly-resampler figures
  => serialize the plotly-resampler figure with its `hf_data` and other plotly-resampler properties, allowig pickling & (deep)copy!
  (the support for pickling allows to start processes that show a plotly-resampler figure via spawning)
* Extend the test-matrix with (1) Windows & macOS, (2) python 3.10

This PR does;
* #92 
  - [x] support figure as a dict as input for our wrappers (e.g. `FigureResampler(fig.to_dict())` with `fig` a `go.Figure`)
    => this is in line with the vanilla plotly behavior (see `BaseFigure` constructor documentation)
  - [x] propagate `_grid_str` in our composable constructors
  - [x] test the above
* #93
  - [x] disable `test_fwr_time_based_data_s`
* add serialization support
  => this allows starting a `multiprocessing.Process` on Windows & Mac-OS (as these use spawning) https://docs.python.org/3/library/multiprocessing.html#contexts-and-start-methods
  (note that this already worked on Unix as for such OS forking is used)
  - [x] support serialization of plotly-resampler figures
  - [x] test serialization when pickling
  - [x] test serialization when (deep)copying 
* #97 
  - [x] :bug: fix that inline show dash has correct height when not set in layout
  - [x] :sparkles: call `show_dash` for `FigureResampler` on `_ipython_display_`
  - [x] add `show_dash_kwargs` to `FigureResampler` constructor 
  - [x] test the above
  - [ ] update docs for `show_dash_kwargs` 
* update tests for (1) Windows & Mac-OS, and (2) Python 3.10
  - [x] fix selenium tests for Windows & Mac-OS (#95)
  - [x] add Python 3.10 (#96)

### CODE REVIEW

**Comment by jvdd** [comment]
Ready for review @jonasvdd

File: plotly_resampler/figure_resampler/figurewidget_resampler.py [hunk 1]
```diff
@@ -56,11 +56,33 @@ def __init__(
...
+            # A base figure object, can be; 
+            # - a base plotly figure: go.Figure or go.FigureWidget
+            # - a plotly-resampler figure: subclass of AbstractFigureAggregator
+            # => we first copy the layout, grid_str and grid ref
             f.layout = figure.layout
+            f._grid_str = figure._grid_str
             f._grid_ref = figure._grid_ref
             f.add_traces(figure.data)
+        elif isinstance(figure, dict) and (
+            "data" in figure or "layout" in figure # or "frames" in figure  # TODO
```
Comment (jonasvdd) [inline plotly_resampler/figure_resampler/figurewidget_resampler.py:69]: TODO?

File: plotly_resampler/figure_resampler/figure_resampler.py [hunk 2]
```diff
@@ -41,26 +41,53 @@ def __init__(
         show_mean_aggregation_size: bool = True,
         convert_traces_kwargs: dict | None = None,
         verbose: bool = False,
+        show_dash_kwargs: dict | None = None,
```
Comment (jonasvdd) [inline plotly_resampler/figure_resampler/figure_resampler.py:44]: No docstring for this argument. Generated docs look like this:
image

File: plotly_resampler/figure_resampler/figure_resampler.py [hunk 3]
```diff
@@ -41,26 +41,53 @@ def __init__(
...
+                f._grid_ref = figure.get("_grid_ref")
+                f.add_traces(figure.get("data"))
+                # `pr_props` is not None when loading a pickled plotly-resampler figure
+                f._pr_props = figure.get("pr_props") 
+                # `f._pr_props`` is an attribute to store properties of a
+                # plotly-resampler figure. This attribute is only used to pass
+                # information to the super() constructor. Once the super constructor is
+                # called, the attribute is removed.
+
+                # f.add_frames(figure.get("frames")) TODO
```
Comment (jonasvdd) [inline plotly_resampler/figure_resampler/figure_resampler.py:84]: TODO?

File: plotly_resampler/figure_resampler/figure_resampler.py [hunk 4]
```diff
@@ -130,6 +157,8 @@ def show_dash(
             See more https://dash.plotly.com/dash-core-components/graph
         **kwargs: dict
             Additional app.run_server() kwargs. e.g.: port
+            Note that these kwargs take precedence over the ones passed to the
```
Comment (jonasvdd) [inline plotly_resampler/figure_resampler/figure_resampler.py:160]: would add a line-break between the inline Note and the *e.g.: port*, as it looks a little too clumped now
image

File: tests/conftest.py [hunk 5]
```diff
@@ -144,7 +160,7 @@ def example_figure() -> FigureResampler:
                 name=f"room {i+1}",
             ),
             hf_x=df_data_pc.index,
-            hf_y=df_data_pc[c],
+            hf_y=df_data_pc[c].astype(np.float32),
```
Comment (jonasvdd) [inline tests/conftest.py:163]: why this casting?

File: tests/conftest.py [hunk 6]
```diff
@@ -216,7 +232,7 @@ def example_figure_fig() -> go.Figure:
             go.Scattergl(
                 name=f"room {i+1}",
                 x=df_data_pc.index,
-                y=df_data_pc[c],
+                y=df_data_pc[c].astype(np.float32),
```
Comment (jonasvdd) [inline tests/conftest.py:235]: why this casting?

File: tests/test_figure_resampler.py [hunk 7]
```diff
@@ -924,6 +942,60 @@ def test_fr_object_binary_data():
...
     assert binary_series_o.dtype == object
     assert len(fig.hf_data) == 1
-    assert fig.hf_data[0]["y"].dtype == "int64"
-    assert fig.data[0]["y"].dtype == "int64"
+    assert (fig.hf_data[0]["y"].dtype == "int32") or (fig.hf_data[0]["y"].dtype == "int64")
+    assert str(fig.data[0]["y"].dtype).startswith("int")
     assert np.all(fig.data[0]["y"] == binary_series)
+
+
+def test_fr_copy_grid():
```
Comment (jonasvdd) [inline tests/test_figure_resampler.py:950]: why are there no tests / checks for _grid_str?

File: tests/test_figurewidget_resampler.py [hunk 8]
```diff
@@ -1793,6 +1813,60 @@ def test_fwr_object_binary_data():
...
     assert binary_series_o.dtype == object
     assert len(fig.hf_data) == 1
-    assert fig.hf_data[0]["y"].dtype == "int64"
-    assert fig.data[0]["y"].dtype == "int64"
+    assert (fig.hf_data[0]["y"].dtype == "int32") or (fig.hf_data[0]["y"].dtype == "int64")
+    assert str(fig.data[0]["y"].dtype).startswith("int")
     assert np.all(fig.data[0]["y"] == binary_series)
+
+
+def test_fwr_copy_grid():
```
Comment (jonasvdd) [inline tests/test_figurewidget_resampler.py:1821]: again why no tests/checks for `_grid_str`

**Review by jonasvdd — COMMENTED** [review]
Amazing work  @jvdd!  👏🏼  💪🏼 🤝, If the comments are resolved, a merge should be fine!

File: plotly_resampler/figure_resampler/figurewidget_resampler.py (hunk 1 above)
Comment (jvdd) [reply plotly_resampler/figure_resampler/figurewidget_resampler.py:69]: Well we did not support it before, nor do we support it now. Hence TODO
(wasn't really aware of frames functionality of plotly)

File: plotly_resampler/figure_resampler/figurewidget_resampler.py [hunk 9]
```diff
@@ -56,11 +56,33 @@ def __init__(
...
+            f._grid_ref = figure.get("_grid_ref")
+            f.add_traces(figure.get("data"))
+            # `pr_props` is not None when loading a pickled plotly-resampler figure
+            f._pr_props = figure.get("pr_props") 
+            # `f._pr_props`` is an attribute to store properties of a plotly-resampler
+            # figure. This attribute is only used to pass information to the super()
+            # constructor. Once the super constructor is called, the attribute is
+            # removed.
+
+            # f.add_frames(figure.get("frames")) TODO
```
[the remark this answers is not in the document]
Comment (jvdd) [reply plotly_resampler/figure_resampler/figurewidget_resampler.py:85]: See above

File: plotly_resampler/figure_resampler/figurewidget_resampler.py (hunk 1 above)
Comment (jonasvdd) [reply plotly_resampler/figure_resampler/figurewidget_resampler.py:69]: should be in FAQ!

File: tests/conftest.py (hunk 5 above)
Comment (jvdd) [reply tests/conftest.py:163]: serialization (smaller size, faster loading / saving)

File: tests/conftest.py (hunk 6 above)
Comment (jvdd) [reply tests/conftest.py:235]: see above

File: tests/test_figure_resampler.py (hunk 7 above)
Comment (jvdd) [reply tests/test_figure_resampler.py:950]: Valid remark! Added these in the latest commit :)

File: tests/test_figurewidget_resampler.py (hunk 8 above)
Comment (jvdd) [reply tests/test_figurewidget_resampler.py:1821]: See above

File: plotly_resampler/figure_resampler/figurewidget_resampler.py [hunk 10]
```diff
@@ -56,11 +56,33 @@ def __init__(
         f = self._get_figure_class(go.FigureWidget)()
         f._data_validator.set_uid = False
 
-        if isinstance(figure, BaseFigure):  # go.Figure or go.FigureWidget or AbstractFigureAggregator
-            # A base figure object, we first copy the layout and grid ref
+        if isinstance(figure, BaseFigure):
```
[the remark this answers is not in the document]
Comment (jonasvdd) [reply plotly_resampler/figure_resampler/figurewidget_resampler.py:59]: aha, this is caused by the inheritence of the `FigureWidget` class.

File: plotly_resampler/figure_resampler/figure_resampler.py (hunk 2 above)
Comment (jonasvdd) [reply plotly_resampler/figure_resampler/figure_resampler.py:44]: fixed.

**Review by jonasvdd — APPROVED** [review]

**Pull request merged**